src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py

```diff
@@ -243,29 +253,552 @@ def apply(self, node, options=None):
+                           "mandatory in OpenCL < 2.0).")
+                error = Call.create(check_status,
+                                    [Literal(message, CHARACTER_TYPE),
+                                     Literal("-1", INTEGER_TYPE)])
+                ifblock = IfBlock.create(check, [error])
+                node.children.insert(outerloop.position, ifblock)
+
+            # Retrieve kernel symbol
+            kernelsym = node.symbol_table.lookup_with_tag(
+                            "kernel_" + kern.name)
+
+            # Choose the command queue number to which to dispatch this kernel.
+            # We have do deal with possible dependencies to kernels dispatched
+            # in different command queues as the order of execution is not
+            # guaranteed.
+            queue_number = kern.opencl_options['queue_number']
+            cmd_queue = ArrayReference.create(
+                    qlist, [Literal(str(queue_number), INTEGER_TYPE)])
+            dependency = outerloop.backward_dependence()
+
+            # If the dependency is a loop containing a kernel, add a barrier if
+            # the previous kernels were dispatched in a different command queue
+            if dependency and dependency.coded_kernels():
+                for kernel_dep in dependency.coded_kernels():
+                    previous_queue = kernel_dep.opencl_options['queue_number']
+                    if previous_queue != queue_number:
+                        # If the backward dependency is being executed in
+                        # another queue we add a barrier to make sure the
+                        # previous kernel has finished before this one starts.
+                        barrier = Assignment.create(
+                                    Reference(flag),
+                                    Call.create(cl_finish, [
+                                        ArrayReference.create(qlist, [
+                                            Literal(str(previous_queue),
+                                                    INTEGER_TYPE)])]))
+                        node.children.insert(outerloop.position, barrier)
+
+            # If the dependency is something other than a kernel, currently we
+            # dispatch everything else to queue _OCL_MANAGEMENT_QUEUE, so add a
+            # barrier if this kernel is not on queue _OCL_MANAGEMENT_QUEUE.
+            if dependency and not dependency.coded_kernels() and \
+                    queue_number != self._OCL_MANAGEMENT_QUEUE:
+                barrier = Assignment.create(
+                            Reference(flag),
+                            Call.create(cl_finish, [
+                                ArrayReference.create(qlist, [
+                                    Literal(str(self._OCL_MANAGEMENT_QUEUE),
+                                            INTEGER_TYPE)])]))
+                node.children.insert(outerloop.position, barrier)
+
+            if api_config.debug_mode:
```

Comment:
Another utility function to add this checking code would help reduce the size of this method.

Reply:
Done